compiler/rustc_mir/src/transform/check_unsafety.rs

```diff
@@ -232,37 +232,40 @@ impl<'a, 'tcx> Visitor<'tcx> for UnsafetyChecker<'a, 'tcx> {
                     UnsafetyViolationKind::GeneralAndConstFn,
                     UnsafetyViolationDetails::DerefOfRawPointer,
                 ),
-                ty::Adt(adt, _) => {
-                    if adt.is_union() {
-                        if context == PlaceContext::MutatingUse(MutatingUseContext::Store)
-                            || context == PlaceContext::MutatingUse(MutatingUseContext::Drop)
-                            || context == PlaceContext::MutatingUse(MutatingUseContext::AsmOutput)
-                        {
-                            let elem_ty = match elem {
-                                ProjectionElem::Field(_, ty) => ty,
-                                _ => span_bug!(
-                                    self.source_info.span,
-                                    "non-field projection {:?} from union?",
-                                    place
-                                ),
-                            };
-                            if !elem_ty.is_copy_modulo_regions(
+                ty::Adt(adt, _) if adt.is_union() => {
+                    if context == PlaceContext::MutatingUse(MutatingUseContext::Store)
+                        || context == PlaceContext::MutatingUse(MutatingUseContext::Drop)
+                        || context == PlaceContext::MutatingUse(MutatingUseContext::AsmOutput)
+                    {
+                        let elem_ty = match elem {
+                            ProjectionElem::Field(_, ty) => ty,
+                            _ => span_bug!(
+                                self.source_info.span,
+                                "non-field projection {:?} from union?",
+                                place
+                            ),
+                        };
+                        let manually_drop = elem_ty
+                            .ty_adt_def()
+                            .map_or(false, |adt_def| adt_def.is_manually_drop());
```

Comment:
This leads to the funny situation where, inside `core::mem::manually_drop`, unions are unsound, because we can do things like
```rust
union U {
  f: ManuallyDrop<Vec<i32>>,
}

fn test(u: U) {
  u.f.value = Vec::new();
}
```
It's a bit awkward to be checking the type of the field here when really we care about the type of the assignment.

Reply:
Fixed this by checking the actually assigned type, not the type of the projection.